src/main/java/se/bjurr/gitchangelog/internal/git/GitSubmoduleParser.java

```diff
@@ -0,0 +1,110 @@
+package se.bjurr.gitchangelog.internal.git;
+
+import static org.slf4j.LoggerFactory.getLogger;
+
+import com.google.common.base.Optional;
+import java.util.*;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+import org.slf4j.Logger;
+import se.bjurr.gitchangelog.api.GitChangelogApi;
+import se.bjurr.gitchangelog.api.exceptions.GitChangelogRepositoryException;
+import se.bjurr.gitchangelog.api.model.Changelog;
+import se.bjurr.gitchangelog.internal.git.model.GitCommit;
+import se.bjurr.gitchangelog.internal.settings.Settings;
+
+public class GitSubmoduleParser {
+  private static final Logger LOG = getLogger(GitSubmoduleParser.class);
+
+  public GitSubmoduleParser() {}
+
+  public List<Changelog> parseForSubmodules(
+      final GitChangelogApi gitChangelogApi,
+      final boolean useIntegrationIfConfigured,
+      final GitRepo gitRepo,
+      final List<GitCommit> commits) {
+
+    List<Changelog> submodules = new ArrayList<>();
+    Map<String, SubmoduleEntry> submoduleEntries = new TreeMap<>();
+    Pattern submoduleNamePattern =
+        Pattern.compile(
+            "(?m)^\\+{3} b/([\\w/\\s-]+)($\\n@.+)?\\n-Subproject commit (\\w+)$\\n\\+Subproject commit (\\w+)$");
+
+    Settings settings = gitChangelogApi.getSettings();
+
+    Optional<String> cachedFromCommit = settings.getFromCommit();
+    Optional<String> cachedToCommit = settings.getToCommit();
+    Optional<String> cachedFromRef = settings.getFromRef();
+    Optional<String> cachedToRef = settings.getToRef();
+    String cachedFromRepo = settings.getFromRepo();
+
+    for (GitCommit commit : commits) {
+      String diff = gitRepo.getDiff(commit.getHash());
+      Matcher submoduleMatch = submoduleNamePattern.matcher(diff);
+      while (submoduleMatch.find()) {
+        String submoduleName = submoduleMatch.group(1);
+        String previousSubmoduleHash = submoduleMatch.group(3);
+        String currentSubmoduleHash = submoduleMatch.group(4);
+        GitRepo submodule = gitRepo.getSubmodule(submoduleName);
+
+        if (submodule == null) {
+          continue;
+        }
+
+        SubmoduleEntry submoduleEntry =
+            new SubmoduleEntry(
+                submoduleName, previousSubmoduleHash, currentSubmoduleHash, submodule);
+
+        if (!submoduleEntries.containsKey(submoduleName)) {
+          submoduleEntries.put(submoduleName, submoduleEntry);
+        }
+
+        SubmoduleEntry existingEntry = submoduleEntries.getOrDefault(submoduleName, submoduleEntry);
+        existingEntry.previousSubmoduleHash = submoduleEntry.previousSubmoduleHash;
+      }
+    }
+
+    for (Map.Entry<String, SubmoduleEntry> submoduleEntry : submoduleEntries.entrySet()) {
+      settings.setFromCommit(submoduleEntry.getValue().previousSubmoduleHash);
```

Comment:
I dont like changing the settings object, and later change it back. I would rather like a new instance created and used.